src/System.Diagnostics.Tracing.Telemetry/src/System/Diagnostics/Tracing/TelemetryListener.cs

```diff
@@ -0,0 +1,249 @@
+                m_next = s_allListeners;
+                s_allListeners = this;
+            }
+        }
+
+        /// <summary>
+        /// Clean up the NotificationListeners.   Notification listeners do NOT DIE ON THEIR OWN
+        /// because they are in a global list (for discoverability).  You must dispose them explicitly. 
+        /// Note that we do not do the Dispose(bool) pattern because we frankly don't want to support
+        /// subclasses that have non-managed state.   
+        /// </summary>
+        virtual public void Dispose()
+        {
+            // Remove myself from the list if all listeners.  
+            lock (DefaultListener)
+            {
+                m_disposed = true;
+                if (s_allListeners == this)
+                    s_allListeners = s_allListeners.m_next;
+                else
+                {
+                    var cur = s_allListeners;
+                    while (cur != null)
+                    {
+                        if (cur.m_next == this)
+                        {
+                            cur.m_next = this.m_next;
+                            break;
+                        }
+                        cur = cur.m_next;
+                    }
+                }
+                m_next = null;
+            }
+
+            // TODO audit for races...  Should this be under the lock above?  
+
+            // Indicate completion to all subscribers.  
+            Subscription subscriber = null;
+            Interlocked.Exchange(ref subscriber, m_subscriptions);
+            while (subscriber != null)
+            {
+                subscriber.Observer.OnCompleted();
+                subscriber = subscriber.Next;
+            }
+            // The code above also nulled out all subscriptions. 
+            // TODO: should we block new subscriptions after we have been disposed?  
+        }
+
+        /// <summary>
+        /// Return the name for the ToString() to aid in debugging.  
+        /// </summary>
+        /// <returns></returns>
+        public override string ToString()
+        {
+            return Name;
+        }
+
+        #region private
+
+        // NotificationSource implementation
+        public override bool IsEnabled(string telemetryName)
+        {
+            for (var curSubscription = m_subscriptions; curSubscription != null; curSubscription = curSubscription.Next)
+            {
+                if (curSubscription.IsEnabled == null || curSubscription.IsEnabled(telemetryName))
+                    return true;
+            }
+            return false;
+        }
+
+        public override void WriteTelemetry(string telemetryName, object parameters)
+        {
+            for (var curSubscription = m_subscriptions; curSubscription != null; curSubscription = curSubscription.Next)
+                curSubscription.Observer.OnNext(new KeyValuePair<string, object>(telemetryName, parameters));
```

Comment:
Same question as for the OnCompleted earlier: what's the intended behavior if OnNext throws?  Do you want to continue notifying other subscribers, or stop notifying and allow the exception to propagate?